Savory Magenta Peacock

medium

# OptimismPortal2:finalizeWithdrawalTransactionExternalProof wrongly checks callWithMinGas success result

## Summary
The finalizeWithdrawalTransactionExternalProof is the last step towards cross-chain message/withdrawal execution at the OptimismPortal2 contract.
It does set the withdrawalHash value to true at the finalizedWithdrawals mapping then calls the intended transaction with a SafeCall.callWithMinGas call.
The issue is the success boolean returned value from this call is not properly handled, leaving failing calls to never be retried.

## Vulnerability Detail
Take a look at the following code snippet:
```solidity
function finalizeWithdrawalTransactionExternalProof(
        Types.WithdrawalTransaction memory _tx,
        address _proofSubmitter
    )
        public
        whenNotPaused
    {
    ...
	    bool success = SafeCall.callWithMinGas(_tx.target, _tx.gasLimit, _tx.value, _tx.data);
        ...
        if (!success && tx.origin == Constants.ESTIMATION_ADDRESS) {
            revert("OptimismPortal: withdrawal failed");
        }
    }
```

The condition to revert is that the call is not successful **and** the tx.origin is the ESTIMATION_ADDRESS. This is incorrect, as both of the checked conditions should suffice for a call reversion. 

## Impact
Successful calls originated from the ESTIMATION_ADDRESS will not revert.
Unsuccessful calls will not revert and the withdrawal won't be possible to be repeated as its hash value is set to true at the finalizedWithdrawals mapping.

## Code Snippet
https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L381C8-L383C10

## Tool used

Manual Review

## Recommendation
The correct implementation of the conditional involves an OR statement, instead of an AND one:
```solidity
if (!success || tx.origin == Constants.ESTIMATION_ADDRESS) {
            revert("OptimismPortal: withdrawal failed");
}
```

